pkg/reconciler/broker/trigger.go

```diff
@@ -0,0 +1,423 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package broker
+
+import (
+	"context"
+	"fmt"
+
+	brokerv1beta1 "github.com/google/knative-gcp/pkg/apis/broker/v1beta1"
+	"github.com/google/knative-gcp/pkg/broker/config"
+	triggerreconciler "github.com/google/knative-gcp/pkg/client/injection/reconciler/broker/v1beta1/trigger"
+	gpubsub "github.com/google/knative-gcp/pkg/gclient/pubsub"
+	"github.com/google/knative-gcp/pkg/reconciler"
+	"github.com/google/knative-gcp/pkg/reconciler/broker/resources"
+	"github.com/google/knative-gcp/pkg/utils"
+	"go.uber.org/zap"
+	gstatus "google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/eventing/pkg/apis/eventing/v1alpha1"
+	"knative.dev/eventing/pkg/duck"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/logging"
+	pkgreconciler "knative.dev/pkg/reconciler"
+	"knative.dev/pkg/resolver"
+)
+
+const (
+	// Name of the corev1.Events emitted from the Trigger reconciliation process.
+	triggerReconciled         = "TriggerReconciled"
+	triggerReadinessChanged   = "TriggerReadinessChanged"
+	triggerReconcileFailed    = "TriggerReconcileFailed"
+	triggerUpdateStatusFailed = "TriggerUpdateStatusFailed"
+)
+
+type TriggerReconciler struct {
+	*reconciler.Base
+
+	// Dynamic tracker to track KResources. It tracks the dependency between Triggers and Sources.
+	kresourceTracker duck.ListableTracker
+
+	// Dynamic tracker to track AddressableTypes. It tracks Trigger subscribers.
+	addressableTracker duck.ListableTracker
+	uriResolver        *resolver.URIResolver
+
+	// CreateClientFn is the function used to create the Pub/Sub client that interacts with Pub/Sub.
+	// This is needed so that we can inject a mock client for UTs purposes.
+	CreateClientFn gpubsub.CreateFn
+}
+
+// Check that TriggerReconciler implements Interface
+var _ triggerreconciler.Interface = (*TriggerReconciler)(nil)
+var _ triggerreconciler.Finalizer = (*TriggerReconciler)(nil)
+
+func triggerNewReconciledNormal(namespace, name string) pkgreconciler.Event {
+	return pkgreconciler.NewEvent(corev1.EventTypeNormal, triggerReconciled, "Trigger reconciled: \"%s/%s\"", namespace, name)
+}
+
+func (r *TriggerReconciler) ReconcileKind(ctx context.Context, t *brokerv1beta1.Trigger) pkgreconciler.Event {
+	b := brokerFromContext(ctx)
```

Comment:
Would it be easier to directly reconcile triggers (vs. using generated trigger reconciler) like what we have in eventing core?

Reply:
You can bulk upsert targets to the TargetsConfig and signal a single configmap update.

Reply:
The reconcile code is similar. These triggers have finalizers and the core ones don't, so there's some extra complexity. Using the generated trigger reconciler means we have a generated status update and finalizer update, so less code to maintain here. I don't have a strong preference though.

Bulk upsert isn't necessary because configmap update is triggered at the end of the broker reconcile and executed in a singleton goroutine separate from the reconciler worker.